File: B/1382-balance-a-binary-search-tree/main.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func balanceBST(root *TreeNode) *TreeNode {
	buf := []int{}

	var ctor func(root *TreeNode, buf *[]int)
	ctor = func(root *TreeNode, buf *[]int) {
		if root == nil {
			return
		}
		if root.Left != nil {
			ctor(root.Left, buf)
		}
		*buf = append(*buf, root.Val)
		if root.Right != nil {
			ctor(root.Right, buf)
		}
	}

	ctor(root, &buf)

	var build func(buf []int, s, e int) *TreeNode
	build = func(buf []int, s, e int) *TreeNode {
		if s > e {
			return nil
		}
		m := s + (e-s)/2
		tn := TreeNode{Val: buf[m]}
		tn.Left = build(buf, s, m-1)
		tn.Right = build(buf, m+1, e)
		return &tn
	}

	return build(buf, 0, len(buf)-1)
}

func main() {
	tn := TreeNode{Val: 2, Left: &TreeNode{Val: 1}, Right: &TreeNode{Val: 3, Right: &TreeNode{Val: 4}}}
	balanceBST(&tn)
}

File: B/1382-balance-a-binary-search-tree/main.rs
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
  pub fn balance_bst(root: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
    fn collect(node: Option<Rc<RefCell<TreeNode>>>, coll: &mut Vec<i32>) {
      if None == node {
        return;
      }
      let node = node.unwrap();
      collect(node.borrow().left.clone(), coll);
      coll.push(node.borrow().val);
      collect(node.borrow().right.clone(), coll);
    }

    fn setup(coll: &Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
      if coll.len() == 0 {
        return None;
      }

      let m = coll.len() / 2;
      Some(Rc::new(RefCell::new(TreeNode {
        val: coll[m],
        left: setup(&coll[0..m].to_vec()),
        right: setup(&coll[m + 1..coll.len()].to_vec()),
      })))
    }
    let mut coll: Vec<i32> = vec![];
    collect(root, &mut coll);
    setup(&coll)
  }
}

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None,
    }
  }
}
